### PR TITLE
add support for huggingface transformers object detectors

### LINKED ISSUE #381: add huggingface detector support
add [DetrForObjectDetection](https://huggingface.co/docs/transformers/v4.16.2/en/model_doc/detr#transformers.DetrForObjectDetection) support by utilizing [AutoModelForObjectDetection](https://huggingface.co/docs/transformers/v4.16.2/en/model_doc/auto#transformers.AutoModelForObjectDetection)

### PR DESCRIPTION
**Pull request by devrimcavusoglu**
Closes #381.

### CODE REVIEW

**Comment by fcakyon** [comment]
@devrimcavusoglu great PR! I guess it requires reformating to pass the tests 👍

File: tests/test_huggingfacemodel.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+# OBSS SAHI Tool
+# Code written by Fatih C Akyon, 2020.
```
Comment (fcakyon) [inline tests/test_huggingfacemodel.py:2]:
```suggestion
# Code written by Devrim Cavusoglu, 2022.
```

File: tests/test_huggingfacemodel.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            model=huggingface_model,
+            confidence_threshold=CONFIDENCE_THRESHOLD,
+            device=MODEL_DEVICE,
+            category_remapping=None,
+            load_at_init=True,
+        )
+
+        self.assertNotEqual(yolov5_detection_model.model, None)
+
+    def test_perform_inference(self):
```
Comment (fcakyon) [inline tests/test_huggingfacemodel.py:47]: Maybe we can an additional test as `test_automodel_prediction` to test HuggingfaceDetectionModel inference via AutoModel class.

File: sahi/model.py [hunk 3]
```diff
@@ -702,3 +704,192 @@ def _create_object_prediction_list_from_original_predictions(
...
+        except ImportError:
+            raise ImportError(
+                "Please install transformers. Check "
+                "`https://github.com/huggingface/transformers#installation` "
+                "for instalattion details."
+            )
+
+        self.set_model(self.model_path)
+
+    def set_model(self, model: Any):
```
Comment (fcakyon) [inline sahi/model.py:758]: this method should take objects instead of str paths

File: sahi/model.py [hunk 4]
```diff
@@ -702,3 +704,192 @@ def _create_object_prediction_list_from_original_predictions(
...
+        if not isinstance(model, str):
```
Comment (fcakyon) [inline sahi/model.py:759]: `set_model` should take objects instead of str paths

File: sahi/model.py [hunk 5]
```diff
@@ -702,3 +704,192 @@ def _create_object_prediction_list_from_original_predictions(
...
+            raise ValueError("`model` has to be path to the model as feature_extractor is also needed.")
+
+        from transformers import AutoFeatureExtractor, AutoModelForObjectDetection
+
+        self.model = AutoModelForObjectDetection.from_pretrained(self.model_path)
```
Comment (fcakyon) [inline sahi/model.py:764]: these steps should be moved to `load_model`

File: tests/test_huggingfacemodel.py [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+from sahi.utils.cv import read_image
+
+MODEL_DEVICE = "cpu"
+CONFIDENCE_THRESHOLD = 0.3
+IMAGE_SIZE = 320
+TEST_MODEL_PATH = "hustvl/yolos-tiny"
+
+
+class TestYolov5DetectionModel(unittest.TestCase):
```
Comment (fcakyon) [inline tests/test_huggingfacemodel.py:16]:
```suggestion
class TestHuggingfaceDetectionModel(unittest.TestCase):
```

File: tests/test_huggingfacemodel.py (hunk 2 above)
Comment (devrimcavusoglu) [reply tests/test_huggingfacemodel.py:49]: Automodel is only a factory class for loading the desired detection model, so I didn't quite get it how we can test it. Can you give an example ?

Comment (fcakyon) [reply tests/test_huggingfacemodel.py:49]: What I am talking about is something like this: https://github.com/obss/sahi/blob/5320f48a1c8f1b5e4d2168c44ae3fb7e94511139/tests/test_predict.py#L121

You can consider it as an integration test.

It is not compulsory though, we can skip it.

Comment (devrimcavusoglu) [reply tests/test_huggingfacemodel.py:49]: @fcakyon Oh okay, I can add it. I've mistaken as transformers' automodel.

Comment (devrimcavusoglu) [reply tests/test_huggingfacemodel.py:51]: @fcakyon all tests are added in `test_predict.py` for HF model.

**Review by fcakyon — APPROVED** [review]

**Pull request merged**